### PR TITLE
 Correctly deallocate prepared statements if we fail inside a transaction

### PR DESCRIPTION
**Pull request by samsondav**
- Addresses issue #12330
# Overview

Cached postgres prepared statements become invalidated if the schema
changes in a way that it affects the returned result.

Examples:
- adding or removing a column then doing a 'SELECT *'
- removing the foo column  then doing a 'SELECT bar.foo'

In normal operation this isn't a problem, we can rescue the error,
deallocate the prepared statement and re-issue the command.

However in PostgreSQL transactions, once any command fails, the
transaction becomes 'poisoned' and any subsequent commands will raise
InFailedSQLTransaction.

This includes DEALLOCATE statements, so the default deallocation
strategy instead of removing the cached prepared statement instead
raises InFailedSQLTransaction.
# Why this is bad
1. InFailedSQLTransaction is a fairly cryptic error and doesn't
   communicate any useful information about what has actually gone wrong.
2. In the naive implementation the prepared statement never gets
   deallocated - it stays alive for the length of the session taking up
   memory on the postgres server.
3. It is unsafe to retry the transaction because the bad prepared
   statement is still in the cache and we would see the exact same failure
   repeated.
# Solution

If we are outside a transaction we can continue to handle these failures
gracefully in the usual way.

Inside a transaction instead of issuing a DEALLOCATE command that will
certainly fail, we now raise
ActiveRecord::PreparedStatementCacheExpired.

This can be handled further up the stack, notably inside
TransactionManager#within_new_transaction. Here we can make sure to
first rollback the transaction, then safely issue DEALLOCATE statements
to invalidate the rest of the cached prepared statements.

This also allows the user (or some gem) the opportunity to catch this error and
voluntarily retry the transaction if a schema change causes the prepared
statement cache to become invalidated.

Because the outdated statement has been deallocated, we can expect the
transaction to succeed on the second try.

### CODE REVIEW

**Comment by samphilipd** [comment]
@sgrif Going to look into the test failures on this one, will ping you when they are resolved.

**Comment by samphilipd** [comment]
@sgrif OK tests passing and I cleanup up a few things, this is ready for review.

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 1]
```diff
@@ -589,25 +589,42 @@ def exec_cache(sql, name, binds)
...
-          # prepared statements whose return value may have changed is
-          # FEATURE_NOT_SUPPORTED.  Check here for more details:
-          # http://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/backend/utils/cache/plancache.c#l573
-          begin
-            code = pgerror.result.result_error_field(PGresult::PG_DIAG_SQLSTATE)
-          rescue
-            raise e
-          end
-          if FEATURE_NOT_SUPPORTED == code
+          raise unless is_cached_plan_failure?(e)
```
Comment (HoyaBoya) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Out of curiosity, would this be ```raise e```? 

I'm tracking this PR for our own issues with prepared statements in Rails 4-2. This is really awesome work.

Comment (samphilipd) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: In this particular case, `raise e` and `raise` have identical behavior. When called without an argument, `raise` simply re-raises the last exception, or a `RuntimeError` if there isn't one.

See http://ruby-doc.org/core-2.1.1/Kernel.html#method-i-raise .

**Comment by samphilipd** [comment]
I also considered adding a `after_failed_transaction_hooks` array to the transaction manager, but this turned out to be too general for just this one specific case.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 2]
```diff
@@ -208,7 +211,16 @@ def current_transaction
       end
 
       private
+
         NULL_TRANSACTION = NullTransaction.new
+
+        # Deallocate invalidated prepared statements outside of the transaction
+        def after_failure_actions(transaction, error)
+          return unless transaction.is_a?(RealTransaction)
+          return unless error.is_a?(ActiveRecord::PreparedStatementCacheExpired)
```
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: if this is a Postgress specific error, I am not sure if we should have handled by all adapters.

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 3]
```diff
@@ -589,25 +589,41 @@ def exec_cache(sql, name, binds)
...
-            raise e
-          end
-          if FEATURE_NOT_SUPPORTED == code
+          # Nothing we can do if we are in a transaction because all commands
+          # will raise InFailedSQLTransaction
+          if in_transaction?
+            raise ActiveRecord::PreparedStatementCacheExpired.new(e.cause.message)
+          else
+            # outside of transactions we can simply flush this query and retry
             @statements.delete sql_key(sql)
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Should we just kill the whole cache here too?

We got lucky on this query, but assuming we have other plans referencing the invalidated relation, the next query we hit could easily be inside a transaction.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 2 above)
Comment (samphilipd) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: @arthurnn I agree but this is the lowest possible level of abstraction I could find that let us actually execute queries outside of the failed transaction.

How would you propose to implement this only for postgres?

Comment (samphilipd) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: I agree in principle, except I couldn't find any place to put adapter-specific transaction behavior.

This method is the 'lowest' possible point in which we can access something outside of the transaction. While it's not ideal, I can't see a less invasive way of implementing this.

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb (hunk 3 above)
Comment (samphilipd) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: @matthewd I deliberately left this behavior exactly as it was before to make this change as non-invasive as possible whilst still solving the core issue.

I'm not an expert here, the person to ask would be @tenderlove who wrote this original case.

I think that invalidating the entire cache in all cases would be a safe response. The only time this Postgres exception is likely to be raised is on a schema change. These are rare enough that the performance hit of flushing the entire prepared statement cache is a minor issue.

**Comment by mecampbellsoup** [comment]
@samphilipd great work on this PR and the issue it originated from - saves us a lot of time and headache, shout out to you! 😄 

I wanted to ask: do you have any patterns you'd recommend for handling the `ActiveRecord::PreparedStatementCacheExpired` exceptions which are now generated? I'm trying to determine where to rescue the exception - e.g. as far upstream as `ApplicationController`? Maybe in `ApplicationRecord` callback/hook so that all the models communicating with the DB can be programmed to auto-retry the Postgres DB command when they see this exception?

**Comment by samphilipd** [comment]
@mecampbellsoup if you are **absolutely sure** that all your transactions are free from side-effects, you could do something like this:

```ruby
class ApplicationRecord
  def transaction(*args, &block)
    retried = false
    super(*args, &block)
  rescue ActiveRecord::PreparedStatementCacheExpired
    if retried
      raise
    else
      retry
    end
  end
end
```

However, if you are sending emails or doing other such things that interact with the outside world inside your transactions, this could result in some of those things maybe happening twice. You should prepare for that case by trying to make your transactions side-effect free.

**Comment by samphilipd** [comment]
@mecampbellsoup here is [a more detailed article](http://samueldavies.net/2017/05/06/how-to-automatically-retry-transactions-instead-of-raising-on-activerecord-preparedstatementcacheexpired/) on how to handle these errors.

**Comment by marisveide** [comment]
I am on Rails 5.1.6 and still are getting this error when underlying schema changes while the Rails process is running.
Just want to make sure - is this fix actually released?

Thanks!

**Comment by bf4** [comment]
@marisveide there are a number of ways to check if a commit is present in a given branch using a version control system like git.  They probably all are quicker for you to do than the time everyone on this issue takes to read your question.

I'm just going to paste the link the merge commit github provides up on the page since it references this commit being present from Rails 5.0 beta through Rails 6 https://github.com/rails/rails/commit/833b14c8f8cae2f2ad2ea31669380b51e644c7db

**Comment by marisveide** [comment]
Hi, @bf4 !
Thank you for taking your time to reply.
The issue is that we are getting this issue on Rails 5.1.6. The solution which helped was to switch off the _prepared statements_ - by setting the `prepared_statements: false` in database.yml.
That results in higher DB load, but is not causing the `ActiveRecord::PreparedStatementCacheExpired` exception when altering the DB schema, and consequently - all those workers die.

We have long-running workers on Rails (with Sidekiq) and are getting this error in all workers which were running while DB schema was modified.

Just in case, here is the stack trace from that exception:
```
Error: ActiveRecord::PreparedStatementCacheExpired - ERROR:  cached plan must not change result type

/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/postgresql_adapter.rb:636:in `rescue in exec_cache'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/postgresql_adapter.rb:621:in `exec_cache'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/postgresql_adapter.rb:605:in `execute_and_clear'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/postgresql/database_statements.rb:79:in `exec_query'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/abstract/database_statements.rb:375:in `select_prepared'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/abstract/database_statements.rb:40:in `select_all'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/abstract/query_cache.rb:97:in `select_all'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/querying.rb:39:in `find_by_sql'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/statement_cache.rb:107:in `execute'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/associations/singular_association.rb:50:in `find_target'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/associations/association.rb:157:in `load_target'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/associations/association.rb:53:in `reload'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/associations/singular_association.rb:7:in `reader'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/associations/builder/association.rb:111:in `shop'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activemodel-5.1.6/lib/active_model/validator.rb:148:in `block in validate'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activemodel-5.1.6/lib/active_model/validator.rb:147:in `each'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activemodel-5.1.6/lib/active_model/validator.rb:147:in `validate'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:413:in `block in make_lambda'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:197:in `block (2 levels) in halting'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:601:in `block (2 levels) in default_terminator'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:600:in `catch'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:600:in `block in default_terminator'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:198:in `block in halting'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:507:in `block in invoke_before'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:507:in `each'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:507:in `invoke_before'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:130:in `run_callbacks'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:827:in `_run_validate_callbacks'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activemodel-5.1.6/lib/active_model/validations.rb:405:in `run_validations!'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activemodel-5.1.6/lib/active_model/validations/callbacks.rb:114:in `block in run_validations!'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:131:in `run_callbacks'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/callbacks.rb:827:in `_run_validation_callbacks'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activemodel-5.1.6/lib/active_model/validations/callbacks.rb:114:in `run_validations!'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activemodel-5.1.6/lib/active_model/validations.rb:335:in `valid?'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/validations.rb:65:in `valid?'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/validations.rb:82:in `perform_validations'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/validations.rb:50:in `save!'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/attribute_methods/dirty.rb:43:in `save!'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/transactions.rb:313:in `block in save!'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/transactions.rb:384:in `block in with_transaction_returning_status'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/abstract/database_statements.rb:235:in `block in transaction'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/abstract/transaction.rb:194:in `block in within_new_transaction'
/opt/rubies/ruby-2.4.4/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/abstract/transaction.rb:191:in `within_new_transaction'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/abstract/database_statements.rb:235:in `transaction'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/transactions.rb:210:in `transaction'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/transactions.rb:381:in `with_transaction_returning_status'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/transactions.rb:313:in `save!'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.6/lib/active_record/suppressor.rb:46:in `save!'
/var/app/current/app/models/job.rb:291:in `set_step'
/var/app/current/app/workers/import/worker.rb:150:in `save_file'
/var/app/current/app/workers/import/worker.rb:144:in `do_import'
/var/app/current/app/workers/import/worker.rb:25:in `run_job'
/var/app/current/app/workers/base_worker.rb:12:in `block in perform'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/spawnling-2.1.6/lib/spawnling.rb:122:in `block in run'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/spawnling-2.1.6/lib/spawnling.rb:184:in `block in fork_it'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/spawnling-2.1.6/lib/spawnling.rb:160:in `fork'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/spawnling-2.1.6/lib/spawnling.rb:160:in `fork_it'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/spawnling-2.1.6/lib/spawnling.rb:122:in `run'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/spawnling-2.1.6/lib/spawnling.rb:102:in `initialize'
/var/app/current/app/workers/base_worker.rb:6:in `new'
/var/app/current/app/workers/base_worker.rb:6:in `perform'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:185:in `execute_job'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:167:in `block (2 levels) in process'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/middleware/chain.rb:128:in `block in invoke'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/middleware/chain.rb:133:in `invoke'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:166:in `block in process'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:137:in `block (6 levels) in dispatch'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/job_retry.rb:98:in `local'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:136:in `block (5 levels) in dispatch'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/rails.rb:42:in `block in call'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/execution_wrapper.rb:85:in `wrap'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/reloader.rb:68:in `block in wrap'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/execution_wrapper.rb:85:in `wrap'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/activesupport-5.1.6/lib/active_support/reloader.rb:67:in `wrap'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/rails.rb:41:in `call'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:132:in `block (4 levels) in dispatch'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:243:in `stats'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:127:in `block (3 levels) in dispatch'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/job_logger.rb:8:in `call'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:126:in `block (2 levels) in dispatch'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/job_retry.rb:73:in `global'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:125:in `block in dispatch'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/logging.rb:48:in `with_context'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/logging.rb:42:in `with_job_hash_context'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:124:in `dispatch'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:165:in `process'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:83:in `process_one'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/processor.rb:71:in `run'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/util.rb:16:in `watchdog'
/opt/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/sidekiq-5.2.2/lib/sidekiq/util.rb:25:in `block in safe_thread'
```

Any suggestions on how to proceed with seeking a fix to that?
Is there a chance that this issue is still not fixed, or that it's another one which causes the same exception?

**Comment by brasic** [comment]
@marisveide this PR does not automatically fix that error, because the only way to fix it is to `retry` the entire `transaction` block, which is not safe in the general case since it can result in side effects like API calls or background job enqueues happening more times than you expect.  The PR merely provides a framework for handling the error in a way that will be application-specific.

If you are sure that every rails transaction block in your codebase has no side effects you can follow the advice in [this article](https://samueldavies.net/2017/05/06/how-to-automatically-retry-transactions-instead-of-raising-on-activerecord-preparedstatementcacheexpired/).  Otherwise, you should define a new `ApplicationRecord.retriable_transaction` method modeled after rails' version that rescues `ActiveRecord::PreparedStatementCacheExpired` and retries, then use that method instead of `transaction` for your pure transactions and continue to use `transaction` for nonpure code.

**Pull request merged**